runtime/lua/vim/_editor.lua

```diff
@@ -627,6 +627,113 @@ function vim.notify(msg, level, opts) -- luacheck: no unused args
   end
 end
 
+-- Helper function to decode the UTF-8 character to a Unicode code point
+---@private
+---@param s string
+---@return integer
+local function utf_ptr2char(s)
+  local b1 = s:byte(1)
+
+  if b1 < 0x80 then
+    -- 1-byte sequence (ASCII character)
+    return b1
+  elseif b1 < 0xE0 then
+    -- 2-byte sequence
+    local b2 = s:byte(2)
+    return ((b1 % 0x20) * 0x40) + (b2 % 0x40)
+  elseif b1 < 0xF0 then
+    -- 3-byte sequence
+    local b2 = s:byte(2)
+    local b3 = s:byte(3)
+    return ((b1 % 0x10) * 0x1000) + ((b2 % 0x40) * 0x40) + (b3 % 0x40)
+  else
+    -- 4-byte sequence
+    local b2 = s:byte(2)
+    local b3 = s:byte(3)
+    local b4 = s:byte(4)
+    return ((b1 % 0x08) * 0x40000) + ((b2 % 0x40) * 0x1000) + ((b3 % 0x40) * 0x40) + (b4 % 0x40)
+  end
+end
+
+--- Convert UTF-32 or UTF-16 {index} to byte index.
+--- Opts can be a table containing: a "encoding" property of "utf-8", "utf-16" or "utf-32"
+--- It can also be supplied as a boolean, true for utf-16 and false for utf-32.
+--- If {opts} is not supplied, it defaults to false (UTF-32). Returns the byte index.
+---
+--- Invalid UTF-8 and NUL is treated like in |vim.str_utfindex()|.
+--- An {index} in the middle of a UTF-16 sequence is rounded upwards to
+--- the end of that sequence.
+---@param s string
+---@param index integer
+---@param opts boolean|nil| { encoding: "utf_8"|"utf-16"|"utf-32" }
+---@return integer
+function vim.str_byteindex(s, index, opts)
```

Comment:
even if the interface continues to live on `vim.x()`, perhaps the implementation could live in `runtime/lua/vim/text.lua` ? (as a private/interval function `text.__foo()`)

Reply:
I did look at `vim.text` as a home for this. But didn't want to change the namespace to keep this change non-breaking.

Happy to move it with a wrapper if you think that is better (and will be bytecode compiled still)

Reply:
> `_editor.lua` is guaranteed to be "compiled" in the `nvim` binary as lua byte code. I'm not sure if that applies to text.lua .

for this reason it should probably stay where you have it.